src/pkg/scan/job.go

```diff
@@ -0,0 +1,337 @@
+	"github.com/goharbor/harbor/src/pkg/scan/report"
+	v1 "github.com/goharbor/harbor/src/pkg/scan/rest/v1"
+	"github.com/pkg/errors"
+)
+
+const (
+	// JobParamRegistration ...
+	JobParamRegistration = "registration"
+	// JobParameterRequest ...
+	JobParameterRequest = "scanRequest"
+	// JobParameterMimes ...
+	JobParameterMimes = "mimeTypes"
+
+	checkTimeout       = 30 * time.Minute
+	firstCheckInterval = 2 * time.Second
+)
+
+// CheckInReport defines model for checking in the scan report with specified mime.
+type CheckInReport struct {
+	Digest           string `json:"digest"`
+	RegistrationUUID string `json:"registration_uuid"`
+	MimeType         string `json:"mime_type"`
+	RawReport        string `json:"raw_report"`
+}
+
+// FromJSON parse json to CheckInReport
+func (cir *CheckInReport) FromJSON(jsonData string) error {
+	if len(jsonData) == 0 {
+		return errors.New("empty JSON data")
+	}
+
+	return json.Unmarshal([]byte(jsonData), cir)
+}
+
+// ToJSON marshal CheckInReport to JSON
+func (cir *CheckInReport) ToJSON() (string, error) {
+	jsonData, err := json.Marshal(cir)
+	if err != nil {
+		return "", errors.Wrap(err, "To JSON: CheckInReport")
+	}
+
+	return string(jsonData), nil
+}
+
+// Job for running scan in the job service with async way
+type Job struct{}
+
+// MaxFails for defining the number of retries
+func (j *Job) MaxFails() uint {
+	return 3
+}
+
+// ShouldRetry indicates if the job should be retried
+func (j *Job) ShouldRetry() bool {
+	return true
+}
+
+// Validate the parameters of this job
+func (j *Job) Validate(params job.Parameters) error {
+	if params == nil {
+		// Params are required
+		return errors.New("missing parameter of scan job")
+	}
+
+	if _, err := extractRegistration(params); err != nil {
+		return errors.Wrap(err, "job validate")
+	}
+
+	if _, err := extractScanReq(params); err != nil {
+		return errors.Wrap(err, "job validate")
+	}
+
+	if _, err := extractMimeTypes(params); err != nil {
+		return errors.Wrap(err, "job validate")
+	}
+
+	return nil
+}
+
+// Run the job
+func (j *Job) Run(ctx job.Context, params job.Parameters) error {
+	// Get logger
+	myLogger := ctx.GetLogger()
+
+	// Ignore errors as they have been validated already
+	r, _ := extractRegistration(params)
+	req, _ := extractScanReq(params)
+	mimes, _ := extractMimeTypes(params)
+
+	// Print related infos to log
+	printJSONParameter(JobParamRegistration, params[JobParamRegistration].(string), myLogger)
+	printJSONParameter(JobParameterRequest, params[JobParameterRequest].(string), myLogger)
+
+	// Submit scan request to the scanner adapter
+	client, err := v1.DefaultClientPool.Get(r)
+	if err != nil {
+		return errors.Wrap(err, "run scan job")
+	}
+
+	resp, err := client.SubmitScan(req)
+	if err != nil {
+		return errors.Wrap(err, "run scan job")
+	}
+
+	// Exit signal
+	exit := make(chan bool, 1)
+	// Done signal
+	done := make(chan bool, 1)
+	// Collect errors
+	eChan := make(chan error, 1)
+
+	go func() {
```

Comment:
It's heavy to get the errors from get report goroutines.

Reply:
Agree.